### PR TITLE
Refactor high quality texture import

### PR DESCRIPTION
**Pull request by reduz**
* Only two texture import modes for low/high quality now:
  * S3TC/BPTC
  * ETC2/ASTC
* Makes sense given this is the general preferred and most compatible combination of texture compression formats in most platforms.
* Removed lossy_quality from VRAM texture compression options. It was unused everywhere.
* Added a new "high_quality" option to texture import. When enabled, it uses BPTC/ASTC (BC7/ASTC4x4) instead of S3TC/ETC2 (DXT1-5/ETC2,ETCA).
* Changed MacOS export settings so required texture formats depend on the architecture selected.
* Also did some fixes in ASTC support (mipmaps not supported).

This solves the following problems:

* Makes it simpler to import textures as high quality, without having to worry about the specific format used.
* On Apple, compressed texture format support is very difficult both for the editor and for exporting because, depending on the hardware, its either one or the other. This makes exporting and running the editor on MacOS  "Just Work"(tm).
* As the editor can now run on platforms such as web, Mac OS with Apple Silicion and Android, it should no longer be assumed that S3TC/BPTC is available by default for it.

### CODE REVIEW

File: editor/export/editor_export.cpp [hunk 1]
```diff
@@ -129,10 +129,20 @@ void EditorExport::add_export_preset(const Ref<EditorExportPreset> &p_preset, in
 }
 
 String EditorExportPlatform::test_etc2() const {
-	const bool etc2_supported = GLOBAL_GET("rendering/textures/vram_compression/import_etc2");
+	const bool etc2_supported = GLOBAL_GET("rendering/textures/vram_compression/import_etc2_astc");
 
 	if (!etc2_supported) {
-		return TTR("Target platform requires 'ETC2' texture compression. Enable 'Import Etc 2' in Project Settings.");
+		return TTR("Target platform requires 'ETC2/ASTC' texture compression. Enable 'Import Etc 2 Astc' in Project Settings.");
```
Comment (akien-mga) [inline editor/export/editor_export.cpp:135]: We properly capitalize these acronyms now even in the UI, so this should be 'Import ETC2 ASTC'.
For it to work for ASTC, you need to add it to the remaps:
```diff
diff --git a/editor/editor_property_name_processor.cpp b/editor/editor_property_name_processor.cpp
index 18ba19f5f6..5380fddde2 100644
--- a/editor/editor_property_name_processor.cpp
+++ b/editor/editor_property_name_processor.cpp
@@ -115,6 +115,7 @@ EditorPropertyNameProcessor::EditorPropertyNameProcessor() {
 	capitalize_string_remaps["arm64-v8a"] = "arm64-v8a";
 	capitalize_string_remaps["armeabi-v7a"] = "armeabi-v7a";
 	capitalize_string_remaps["arvr"] = "ARVR";
+	capitalize_string_remaps["astc"] = "ASTC";
 	capitalize_string_remaps["bg"] = "BG";
 	capitalize_string_remaps["bidi"] = "BiDi";
 	capitalize_string_remaps["bp"] = "BP";
```

File: editor/export/editor_export.cpp [hunk 2]
```diff
@@ -129,10 +129,20 @@ void EditorExport::add_export_preset(const Ref<EditorExportPreset> &p_preset, in
...
+	}
+
+	return String();
+}
+
+String EditorExportPlatform::test_bc() const {
+	const bool bc_supported = GLOBAL_GET("rendering/textures/vram_compression/import_s3tc_bptc");
+
+	if (!bc_supported) {
+		return TTR("Target platform requires 'S3TC/BPTC' texture compression. Enable 'Import S3tc Bptc' in Project Settings.");
```
Comment (akien-mga) [inline editor/export/editor_export.cpp:145]: Likewise, this should use 'Import S3TC BPTC' (at least it should appear like this in the UI, if it doesn't it's a bug).

File: editor/import/resource_importer_layered_texture.cpp [hunk 3]
```diff
@@ -123,6 +123,9 @@ bool ResourceImporterLayeredTexture::get_option_visibility(const String &p_path,
 	if (p_option == "compress/lossy_quality" && p_options.has("compress/mode")) {
 		return int(p_options["compress/mode"]) == COMPRESS_LOSSY;
 	}
```
Comment (akien-mga) [inline editor/import/resource_importer_layered_texture.cpp:125]: Didn't you plan to remove lossy quality?

File: core/os/os.cpp [hunk 4]
```diff
@@ -553,6 +553,16 @@ void OS::add_frame_delay(bool p_can_draw) {
...
+OS::PreferredTextureFormat OS::get_preferred_texture_format() const {
+#if defined(__arm__) || defined(__aarch64__) || defined(_M_ARM) || defined(_M_ARM64)
+	return PREFERRED_TEXTURE_FORMAT_ETC2_ASTC; // By rule, ARM hardware uses ETC texture compression.
+#elif defined(__x86_64__) || defined(_M_X64) || defined(i386) || defined(__i386__) || defined(__i386) || defined(_M_IX86)
+	return PREFERRED_TEXTURE_FORMAT_S3TC_BPTC; // By rule, X86 hardware prefers S3TC and derivatives.
+#else
+	return PREFERRED_TEXTURE_FORMAT_S3TC_BPTC; // Override in platform if needed.
+#endif
+}
+
```
Comment (bruvzg) [inline core/os/os.cpp:565]: This probably will need an extra check for Rosetta on macOS, I guess override like:

```c++
OS::PreferredTextureFormat OS_MacOS::get_preferred_texture_format() const {
#if defined(__aarch64__)
    return PREFERRED_TEXTURE_FORMAT_ETC2_ASTC; // By rule, ARM hardware uses ETC texture compression.
#elif defined(__x86_64__)
    int32_t ret = 0;
    size_t ret_size = sizeof(ret);
    if (sysctlbyname("sysctl.proc_translated", &ret, &ret_size, nulptr, 0) == 0) {
        if (ret == 1) {
            return PREFERRED_TEXTURE_FORMAT_ETC2_ASTC; // X86 binary running on ARM hardware.
        }
    }
    return PREFERRED_TEXTURE_FORMAT_S3TC_BPTC; // By rule, X86 hardware prefers S3TC and derivatives.
#endif
}
```

Also, is it true for the other ARM platforms (Android, Windows for ARM)?

File: editor/import/resource_importer_layered_texture.cpp (hunk 3 above)
Comment (reduz) [reply editor/import/resource_importer_layered_texture.cpp:125]: Lossy quality is remvoed for VRAM compression, but WebP still uses it.

File: core/os/os.cpp (hunk 4 above)
Comment (reduz) [reply core/os/os.cpp:565]: For Windows on ARM I suppose its the same thing, it uses Android hardware..

**Comment by clayjohn** [comment]
I'm getting a crash when loading a scene. 

I deleted the ``.godot`` folder before trying to import. Then after crashing the first time I deleted the ``.import`` files as well and I received the same crash (albeit pointing to a different line number in ``ResourceImporterTexture::import``
```
ERROR: Trying to unreference a SafeRefCount which is already zero is wrong and a symptom of it being misused.
Upon a SafeRefCount reaching zero any object whose lifetime is tied to it, as well as the ref count itself, must be destroyed.
Moreover, to guarantee that, no multiple threads should be racing to do the final unreferencing to zero.
   at: _check_unref_sanity (./core/templates/safe_refcount.h:176)


================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.0.beta.custom_build (dc5cdf4de4aa003de9e78780f87d7264aa4f9ec1)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7fd112c42520] (??:0)
[2] StringName::unref() (/home/clayjohn/dev/godot/./core/templates/safe_refcount.h:173)
[3] StringName::~StringName() (/home/clayjohn/dev/godot/./core/string/string_name.h:182)
[4] ResourceImporterTexture::import(String const&, String const&, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, DefaultTypedAllocator<HashMapElement<StringName, Variant> > > const&, List<String, DefaultAllocator>*, List<String, DefaultAllocator>*, Variant*) (/home/clayjohn/dev/godot/editor/import/resource_importer_texture.cpp:454)
[5] EditorFileSystem::_reimport_file(String const&, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, DefaultTypedAllocator<HashMapElement<StringName, Variant> > > const*, String const&) (/home/clayjohn/dev/godot/editor/editor_file_system.cpp:2001)
[6] EditorFileSystem::_reimport_thread(unsigned int, EditorFileSystem::ImportThreadData*) (/home/clayjohn/dev/godot/editor/editor_file_system.cpp:2159)
[7] WorkerThreadPool::GroupUserData<EditorFileSystem, void (EditorFileSystem::*)(unsigned int, EditorFileSystem::ImportThreadData*), EditorFileSystem::ImportThreadData*>::callback_indexed(unsigned int) (/home/clayjohn/dev/godot/./core/object/worker_thread_pool.h:152)
[8] WorkerThreadPool::_process_task(WorkerThreadPool::Task*) (/home/clayjohn/dev/godot/core/object/worker_thread_pool.cpp:72)
[9] WorkerThreadPool::_native_low_priority_thread_function(void*) (/home/clayjohn/dev/godot/core/object/worker_thread_pool.cpp:165)
[10] Thread::callback(Thread*, Thread::Settings const&, void (*)(void*), void*) (/home/clayjohn/dev/godot/core/os/thread.cpp:67)
[11] void std::__invoke_impl<void, void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>(std::__invoke_other, void (*&&)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*&&, Thread::Settings&&, void (*&&)(void*), void*&&) (/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/invoke.h:61)
[12] std::__invoke_result<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>::type std::__invoke<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>(void (*&&)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*&&, Thread::Settings&&, void (*&&)(void*), void*&&) (/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/invoke.h:96)
[13] void std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul>(std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul>) (/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/std_thread.h:252)
[14] std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> >::operator()() (/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/std_thread.h:259)
[15] std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > >::_M_run() (/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/std_thread.h:210)
[16] /home/clayjohn/dev/godot/bin/godot.linuxbsd.editor.dev.x86_64.llvm(+0xb2089f3) [0x5651a18e79f3] (msdf-error-correction.cpp:?)
[17] /lib/x86_64-linux-gnu/libc.so.6(+0x94b43) [0x7fd112c94b43] (??:0)
[18] /lib/x86_64-linux-gnu/libc.so.6(+0x126a00) [0x7fd112d26a00] (??:0)
-- END OF BACKTRACE --
============================
```

**Comment by reduz** [comment]
@clayjohn The crash and error seem unrelated to the PR, which is very strange

**Comment by clayjohn** [comment]
> @clayjohn The crash and error seem unrelated to the PR, which is very strange. This PR forces the reimport of all textures, so it may be related to that?

I'm not sure. I am testing on https://github.com/Rytelier/Godot-4-forest-benchmark and I can reliably reproduce the crash if I delete the ``.godot`` folder, or if import textures from an older version (I tried with Beta 15) and then try to import with this PR. Of note, after crashing, if I restart the editor it finishes importing the rest of the textures. So it seems like the issue could be due to the number of textures in the project?

I also tested with https://github.com/gdquest-demos/godot-4-3d-third-person-controller and it did not crash

**Comment by clayjohn** [comment]
Couple notes: 
1. When compressing Icon.png into ASTC I get the following error
```
astcenc: Encoding image size 64x64 to format ASTC_4x4, with mipmaps.
ERROR: Expected Image data size of 64x64x1 (with 6 mipmaps) = 5488 bytes, got 4096 bytes instead.
   at: initialize_data (core/io/image.cpp:2206)
```
It looks like the image size validator doesn't properly account for ASTC
2. I was able to work around the crash above by setting ``editor/import/use_multiple_threads`` to false. The crash happens more frequently when ETC ASTC is enabled. But it also happens on master with both S3TC and ETC2 enabled
3. The ASTC formats are missing from ``Image::format_names[]``
https://github.com/godotengine/godot/blob/a43db5afa4bbec4772be2f296931a6d44bb4cbb3/core/io/image.cpp#L79-L81
Need to add them to the end:
```
"ASTC_4x4",
"ASTC_4x4_HDR",
"ASTC_8x8",
"ASTC_8x8_HDR",
```

**Comment by clayjohn** [comment]
Here is the error log from running on an android device:
```
01-27 17:26:57.152 30718 30753 E godot   : USER ERROR: Unable to open file: res://.godot/imported/icon.svg-218a8f2b3041327d8a5756f3a245f83b.astc.ctex.
01-27 17:26:57.152 30718 30753 E godot   :    at: _load_data (scene/resources/texture.cpp:846)
01-27 17:26:57.152 30718 30753 E godot   : USER ERROR: Failed loading resource: res://.godot/imported/icon.svg-218a8f2b3041327d8a5756f3a245f83b.astc.ctex. Make sure resources have been imported by opening the project in the editor at least once.
01-27 17:26:57.152 30718 30753 E godot   :    at: _load (core/io/resource_loader.cpp:222)
01-27 17:26:57.152 30718 30753 E godot   : USER ERROR: Failed loading resource: res://icon.svg. Make sure resources have been imported by opening the project in the editor at least once.
01-27 17:26:57.152 30718 30753 E godot   :    at: _load (core/io/resource_loader.cpp:222)
01-27 17:26:57.152 30718 30753 E godot   : USER ERROR: Can't load dependency: res://icon.svg.
01-27 17:26:57.152 30718 30753 E godot   :    at: load (core/io/resource_format_binary.cpp:696)
01-27 17:26:57.152 30718 30753 E godot   : USER ERROR: Failed loading resource: res://.godot/exported/133200997/export-3070c538c03ee49b7677ff960a3f5195-main.scn. Make sure resources have been imported by opening the project in the editor at least once.
01-27 17:26:57.152 30718 30753 E godot   :    at: _load (core/io/resource_loader.cpp:222)
01-27 17:26:57.152 30718 30753 E godot   : USER ERROR: Failed loading scene: res://main.tscn
01-27 17:26:57.152 30718 30753 E godot   :    at: start (main/main.cpp:2921)
01-27 17:26:57.153 30718 30753 E godot   : USER ERROR: Condition "default_certs != nullptr" is true.
01-27 17:26:57.153 30718 30753 E godot   :    at: load_default_certificates (modules/mbedtls/crypto_mbedtls.cpp:302)
```
Running on a Pixel 4. The scene is just the default icon on a Sprite2D taking up most of the screen. The icon is set to use high quality import. and only "ETC ASTC" import is enabled in the project settings
[ASTC-test.zip](https://github.com/godotengine/godot/files/10524261/ASTC-test.zip)

**Comment by reduz** [comment]
@clayjohn Fixed some problems in the ASTC implementation, so hoping it works now. I can't manage to load ASTC on desktop (Godot does not recognize the format for some reason).

**Comment by clayjohn** [comment]
> @clayjohn Fixed some problems in the ASTC implementation, so hoping it works now. I can't manage to load ASTC on desktop (Godot does not recognize the format for some reason).

Same error as before. The file it is referencing exists, its almost like its not getting included in the APK though. Do we have logic that selects what imported resources are included in the APK?
```
01-28 10:35:36.267 11948 11980 E godot   : USER ERROR: Unable to open file: res://.godot/imported/icon.svg-218a8f2b3041327d8a5756f3a245f83b.astc.ctex.
01-28 10:35:36.267 11948 11980 E godot   :    at: _load_data (scene/resources/texture.cpp:846)
01-28 10:35:36.267 11948 11980 E godot   : USER ERROR: Failed loading resource: res://.godot/imported/icon.svg-218a8f2b3041327d8a5756f3a245f83b.astc.ctex. Make sure resources have been imported by opening the project in the editor at least once.
```

**Comment by reduz** [comment]
Oh, my mistake, I forgot to add astc to Android and iOS, will do asap.

**Comment by reduz** [comment]
done

**Comment by clayjohn** [comment]
Tested again, the error is gone and the texture loads. But it is totally incorrect

_Vulkan renderer on android_
vulkan

_OpenGL renderer on android_
opengl

_How it looks on desktop (uses BPTC)_
Screenshot from 2023-01-29 14-13-01

**Comment by reduz** [comment]
Damn it.. :unamused: 
The weird thing is that the texture encodes/decodes fine to ASTC because the preview shows fine on dekstop, but somehow it does not seem to work when loaded?

**Comment by reduz** [comment]
I will have to compile on my Mac M1 to test whats going on..

**Comment by clayjohn** [comment]
> Damn it.. unamused The weird thing is that the texture encodes/decodes fine to ASTC because the preview shows fine on dekstop, but somehow it does not seem to work when loaded?

When I look at the preview on desktop it always reports DXT or BPTC

**Comment by reduz** [comment]
@clayjohn yeah I copied the .ctex file from the .godot/import folder to test but as Vulkan is not supported on my nvidia GPU I can't see it decoded.

**Comment by reduz** [comment]
@clayjohn Okay that took me a couple of hours to figure out, but it should be working well now. Make sure to test, but I tested on AMD (which supports ASTC) and it works fine.

**Review by akien-mga — APPROVED** [review]
Looks good to me. I would merge ASAP to get it properly tested in beta 17, as I expect bugs on different platforms :)

I'll wait to see if @clayjohn can at least confirm it works on his Android device.

**Comment by reduz** [comment]
@akien-mga I think its safe to merge at this point, seems to work well enough for me.

**Comment by clayjohn** [comment]
Just tested on my android device and I'm still getting the same output as before

**Review by clayjohn — APPROVED** [review]
Looks good to me. After re-importing and re-exporting with the newest changes this seems to work on Android both in mobile and gl_compatibility

**Comment by jcostello** [comment]
Dump question but this has a visual impact?

**Comment by clayjohn** [comment]
> Dumb question but this has a visual impact?

Yes, same as switching from s3tc to bptc in current betas. There won't be a visual change in the default case though

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**